usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -1004,15 +1004,28 @@ static CK_RV ep11tok_pkey_update(STDLL_TokData_t *tokdata, SESSION *session,
 
 /**
  * Returns true if the session is ok for creating protected keys, false
- * otherwise. The session must be read/write and not public, because the
- * protected key is a private/secret key attribute.
+ * otherwise. The session must be read/write for token objects, and not public
+ * nor SO for private objects.
  */
-static CK_BBOOL ep11tok_pkey_session_read_write(SESSION *session)
+static CK_BBOOL ep11tok_pkey_session_ok_for_obj(SESSION *session,
+                                                OBJECT *key_obj)
 {
-    if (session->session_info.flags & CKF_RW_SESSION)
-        return CK_TRUE;
+    if (object_is_token_object(key_obj) &&
+        (session->session_info.flags & CKF_RW_SESSION) == 0)
+        return CK_FALSE;
 
-    return CK_FALSE;
+    if (object_is_private(key_obj)) {
+        switch (session->session_info.state) {
```

Comment:
Would this check "object_is_private" also apply to AES keys? Do they also have the CKA_PRIVATE attribute? But yes, we saw that there is no routine like "is_secret_key()" in object.c. Just to mention ...

Reply:
Yes, attribute CKA_PRIVATE is defined by the PKCS#11 standard for all key objects (actually for all storage objects, where keys belong to).

There might be a misunderstanding: CKA_PRIVATE does **not** mean that its a private key in the sense of a private RSA key vs. a RSA public key. CKA_PRIVATE just means that this object is private to the USER pin, and thus is stored encrypted with the token's master key.

Reply:
so even an RSA public key object may have CKA_PRIVATE=TRUE (although it does not make too much sense, since it does not contain any sensitive information).